include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+
+      //@}
+
+      /**
+       * @name Independent variables
+       */
+      //@{
+
+      /**
+       * A set of independent variables $\mathbf{X}$ that differentiation we be
+       * performed with respect to.
+       *
+       * The gradients and hessians of dependent variables will be computed
+       * at these finite values.
+       */
+      mutable std::vector<scalar_type> independent_variable_values;
+
+      /**
+       * A set of sensitive independent variables $\mathbf{X}$ that
+       * differentiation we be performed with respect to.
+       *
+       * The gradients and hessians of dependent variables will be computed
+       * using these configured AD numbers. Note that only reverse-mode AD
+       * requires that the sensitive independent variables be stored.
+       */
+      mutable std::vector<ad_type> independent_variables;
+
+      /**
+       * Registered independent variables that have been manipulated for a given
+       * set of operations.
+       */
+      std::vector<bool> touched_independent_variables;
+
+      /**
+       * Registered independent variables that have been extracted and their
+       * sensitivities marked.
+       */
+      mutable std::vector<bool> touched_marked_independent_variables;
+
+      /**
+       * Reset the boolean vector that indicates which independent variables
+       * we've been manipulating for the current set of operations.
+       */
+      void
+      reset_touched_independent_variables();
+
+      /**
+       * Set the actual value of the independent variable $X_{i}$.
+       *
+       * @param[in] value The value to set the index'd independent variable to.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      set_sensitivity_value(const scalar_type &value, const unsigned int index);
+
+      /**
+       * Initialise an independent variable $X_{i}$ such that subsequent
+       * operations performed with it are tracked.
+       *
+       * @note Care must be taken to mark each independent variable only once.
+       *
+       * @note The order in which the independent variables are marked defines the
+       * order of all future internal operations. They must be manipulated in
+       * the same order as that in which they are first marked. If not then,
+       * for example, ADOL-C won't throw an error, but rather it might complain
+       * nonsensically during later computations or produce garbage results.
+       *
+       * @param[out] out An auto-differentiable number that is ready for use in
+       * computations. The operations that are performed with it are recorded on
+       * the tape and will be considered in the computation of dependent
+       * variable values.
+       * @param[in] index The index in the vector of independent variables.
+       */
+      void
+      mark_independent_variable(ad_type &out, const unsigned int index) const;
```

Comment:
> but why not just return the ad_type thing

There are two reasons for this that I can think of off the top of my head:
1. We want to minimise the number of active variables that we generate before handing things off to the user, so each pass by copy creates some overhead. In this way we can avoid that.
2. I've had some instances of data corruption that I've since avoided by never using a pass by copy. I can't remember how these issues came to be and how pervasive the issue was, but it was so frustrating to track down the issue that I simply made a rule to avoid copying AD numbers.

Reply:
Hm, ok. It's not inconsistent with what we do with vectors, for example, just annoying that we can't do copies with what one would intuitively consider scalars.

Reply:
Well, since this is an internal function it doesn't really matter too much what the interface is. Its better that we pass a minimally polluted set of independent variables (the starting point of all AD computations) to our users. With the benchmarking I've done I get the sense that each superfluous operation adds a non-insignificant overhead to the derivative computations performed later.

Reply:
Fair enough. I had also forgotten that it's an internal interface.